### ISSUE TITLE
Reset texture wrapping mode for cube maps

### ISSUE THREAD

**Issue by lexaknyazev** [KhronosGroup/MoltenVK#2064]
The Vulkan spec says:
> `Cube` images ignore the wrap modes specified in the sampler.

Apparently, this is not the case for Metal on Apple silicon. When the 3D direction vector points to a cube's edge or corner, there is a small chance for the internally-computed texel coordinates to be out-of-bounds of the normalized range. If the wrapping mode happens to be `MTLSamplerAddressModeRepeat`, a pixel from the opposite edge of the 2D face would be selected.

MoltenVK should override all wrapping modes to `MTLSamplerAddressModeClampToEdge` for samplers used with cube maps.

**Comment by alyssarosenzweig** [comment]
I was wondering if this was broken in MVK.. sigh. We workaround this in our GL driver, but this is really tricky to deal with correctly in VK... good luck :crying_cat_face:

**Comment by alyssarosenzweig** [comment]
@lexaknyazev Do you know if there are CTS tests for this? If not, do we need some?

**Comment by lexaknyazev** [comment]
I discovered this with the following GL tests. No idea if they have VK counterparts.
```
dEQP-GLES3.functional.texture.filtering.cube.
    sizes.63x63_nearest
    sizes.63x63_nearest_mipmap_nearest
    sizes.63x63_nearest_mipmap_linear
    combinations.nearest_nearest_repeat_repeat
    combinations.nearest_nearest_repeat_mirror
    combinations.nearest_nearest_mirror_repeat
    combinations.nearest_linear_clamp_repeat
    combinations.nearest_linear_repeat_clamp
    combinations.nearest_linear_repeat_repeat
    combinations.nearest_linear_repeat_mirror
    combinations.nearest_mipmap_nearest_nearest_clamp_repeat
    combinations.nearest_mipmap_nearest_nearest_mirror_repeat
    combinations.nearest_mipmap_nearest_linear_clamp_repeat
    combinations.nearest_mipmap_nearest_linear_repeat_clamp
    combinations.nearest_mipmap_nearest_linear_mirror_repeat
    combinations.nearest_mipmap_linear_linear_clamp_repeat
    combinations.nearest_mipmap_linear_linear_repeat_clamp
    combinations.nearest_mipmap_linear_linear_mirror_repeat
```

**Comment by billhollings** [comment]
> I discovered this with the following GL tests. No idea if they have VK counterparts.

Although there are not identically-named VK CTS tests, there are very similarly-named tests, which presumably cover the same area. 

For MoltenVK, the CTS results for `dEQP-VK.texture.filtering.cube.*` are:

```
  Passed:        406/413 (98.3%)
  Failed:        0/413 (0.0%)
  Not supported: 7/413 (1.7%)
```

We don't specifically address the corner case issues within MoltenVK, so it's possible the VK CTS tests don't cover them the way the GL tests do. Or perhaps Metal handles this now. 

Further triage and testing needs to be done.

**Comment by alyssarosenzweig** [comment]
The VK CTS indeed fails to cover this case the way the GL tests do. Judging by name, tests like `dEQP-VK.texture.filtering.cube.combinations.linear.linear.repeat.repeat` should cover this case. However, the VK CTS actually uses clamp-to-edge samplers for this test, not the repeat samplers it's supposed to, creating the coverage gap. Fixing that in the VK CTS 'should' address this coverage gap and cause the tests to start failing.

**Comment by alyssarosenzweig** [comment]
> The VK CTS indeed fails to cover this case the way the GL tests do. Judging by name, tests like `dEQP-VK.texture.filtering.cube.combinations.linear.linear.repeat.repeat` should cover this case. However, the VK CTS actually uses clamp-to-edge samplers for this test, not the repeat samplers it's supposed to, creating the coverage gap. Fixing that in the VK CTS 'should' address this coverage gap and cause the tests to start failing.

...although even fixing this, the tests are still passing for me, so I think there's an underlying VK CTS problem. (The cube map issue is hardware and definitely hit in GL.)

**Comment by alyssarosenzweig** [comment]
> > > The VK CTS indeed fails to cover this case the way the GL tests do. Judging by name, tests like `dEQP-VK.texture.filtering.cube.combinations.linear.linear.repeat.repeat` should cover this case. However, the VK CTS actually uses clamp-to-edge samplers for this test, not the repeat samplers it's supposed to, creating the coverage gap. Fixing that in the VK CTS 'should' address this coverage gap and cause the tests to start failing.
> > 
> > 
> > ...although even fixing this, the tests are still passing for me, so I think there's an underlying VK CTS problem. Maybe it's not actually testing the wrapping behaviour?
> 
> Probably not. Without looking at it too deeply (famous last words), I'd venture a guess that the expected output image would probably need to be updated as well. If I'm right, the CTS with your change should fail on _other_ hardware that implements this correctly.

In theory the CTS should be calculating the reference image based on the provided parameters but yeah 🤷‍♀️

**Comment by Tobski** [comment]
@lexaknyazev is this something the gltf viewer is relying upon? I'm quite tempted to just say we should just ban this in core Vulkan rather than requiring this to be worked around, which as @alyssarosenzweig states, would be very hard to do with dynamic indexing, and is likely to hurt performance if nothing else...

**Comment by lexaknyazev** [comment]
I don't think that any glTF viewer should rely on ambiguous cubemap sampling. Disallowing all wrapping modes except clamp-to-edge for cubemaps sounds good to me assuming that VVL and CTS will be appropriately updated.

**Comment by Tobski** [comment]
Thanks, I'll raise an internal MR to see if we can get agreement on this; VVL will get updated, and that should trickle into CTS once validation goes live; but note that there's ongoing work to validate dynamic indexing cases properly, so it'll be a little while until we have full validation.

**Comment by Tobski** [comment]
We discussed this idea in the Vulkan WG and ultimately agreed that given the likelihood that this touches every single app using cubemaps, it's too big of an intrusion to pull into the core spec. HoneyKrisp plans to work around this anyway, and if necessary we can add a portability bit for it (or MoltenVK can do something similar to HoneyKrisp).
